control-plane/grpc/src/operations/volume/traits.rs

```diff
@@ -0,0 +1,1072 @@
+                    Ok(volumeid) => volumeid,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                })
+            }
+        })
+    }
+}
+
+/// Trait to validate the Grpc type by an intermediate conversion
+pub trait ValidateRequestTypes {
+    /// The validated type
+    type Validated;
+    /// The method that is needed to be implemented to ensure validation
+    fn validated(self) -> Result<Self::Validated, ReplyError>;
+}
+
+/// Trait to be implemented for CreateVolume operation
+pub trait CreateVolumeInfo: Send + Sync {
+    /// Uuid of the volume
+    fn uuid(&self) -> VolumeId;
+    /// Size in bytes of the volume
+    fn size(&self) -> u64;
+    /// No of replicas of the volume
+    fn replicas(&self) -> u64;
+    /// Volume policy of the volume, i.e self_heal
+    fn policy(&self) -> VolumePolicy;
+    /// Topology configuration of the volume
+    fn topology(&self) -> Option<Topology>;
+    /// Labels to be added to the volumes for topology based scheduling
+    fn labels(&self) -> Option<VolumeLabels>;
+}
+
+impl CreateVolumeInfo for CreateVolume {
+    fn uuid(&self) -> VolumeId {
+        self.uuid.clone()
+    }
+
+    fn size(&self) -> u64 {
+        self.size
+    }
+
+    fn replicas(&self) -> u64 {
+        self.replicas
+    }
+
+    fn policy(&self) -> VolumePolicy {
+        self.policy.clone()
+    }
+
+    fn topology(&self) -> Option<Topology> {
+        self.topology.clone()
+    }
+
+    fn labels(&self) -> Option<VolumeLabels> {
+        self.labels.clone()
+    }
+}
+
+/// Intermediate structure that validates the conversion to CreateVolumeRequest type
+pub struct ValidatedCreateVolumeRequest {
+    inner: CreateVolumeRequest,
+    uuid: VolumeId,
+    topology: Option<Topology>,
+}
+
+impl CreateVolumeInfo for ValidatedCreateVolumeRequest {
+    fn uuid(&self) -> VolumeId {
+        self.uuid.clone()
+    }
+
+    fn size(&self) -> u64 {
+        self.inner.size
+    }
+
+    fn replicas(&self) -> u64 {
+        self.inner.replicas
+    }
+
+    fn policy(&self) -> VolumePolicy {
+        match self.inner.policy.clone() {
+            Some(policy) => policy.into(),
+            None => VolumePolicy::default(),
+        }
+    }
+
+    fn topology(&self) -> Option<Topology> {
+        self.topology.clone()
+    }
+
+    fn labels(&self) -> Option<VolumeLabels> {
+        match self.inner.labels.clone() {
+            None => None,
+            Some(labels) => Some(labels.value),
+        }
+    }
+}
+
+impl ValidateRequestTypes for CreateVolumeRequest {
+    type Validated = ValidatedCreateVolumeRequest;
+    fn validated(self) -> Result<Self::Validated, ReplyError> {
+        Ok(ValidatedCreateVolumeRequest {
+            uuid: match self.uuid.clone() {
+                Some(uuid) => match VolumeId::try_from(uuid) {
+                    Ok(volumeid) => volumeid,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
```

Comment:
rather than unwrap_err, I think we want to pass the error string into invalid_argument_err